### PR TITLE
Update ODIS SDK to accept any type of identifier

### LINKED ISSUE #9973: Update SDK to allow for all types of identifiers

### PR DESCRIPTION
**Pull request by isabellewei**
### Description

Creates new functions to interact with ODIS that are not phone number specific

Instead of using the phone number specific identifier and hashing logic that was in the `base` and `utils` libraries, I created new functions for the equivalent directly in the odis sdk. I figured that no one would really use that logic outside of the odis sdk anyways (is that a fair assumption?) Eventually I'd like to [remove all phone number specific ODIS code](https://github.com/celo-org/celo-monorepo/issues/9987)

I wanted to use [DID methods](https://w3c.github.io/did-spec-registries/#did-methods) for the prefixes, but seems like as of now there's only one for twitter, but not email or phone number yet. (As an aside, would be cool if we drafted DID methods for them ourselves!)

### Tested

tests updated
backward compatibility test to ensure phone number identifiers are the same

### Related issues

- Fixes #9973 

### Backwards compatibility

- some functions in `phone-number-identifier.ts` are kept to maintain backwards compatibility 

### Documentation

_The set of community facing docs that have been added/modified because of this change_

### CODE REVIEW

File: packages/sdk/identity/src/odis/identifier.ts
Comment (isabellewei) [inline packages/sdk/identity/src/odis/identifier.ts]: this variable used to be a salt in the old code, but it's a little confusing that we suddenly switch to salt when we've been referring to it as a pepper the whole time

File: packages/sdk/identity/src/odis/identifier.ts
Comment (isabellewei) [inline packages/sdk/identity/src/odis/identifier.ts]: do we have any preferences on `offchain`/`onchain` vs `offChain`/`onChain`

File: packages/sdk/identity/src/odis/identifier.ts
Comment (eelanagaraj) [inline packages/sdk/identity/src/odis/identifier.ts]: One concern we discussed at the offsite around this is that without including prefixes in the actual blinded message passed to ODIS (which is IIUC what we're doing here) is that we will get the same pepper for different types of identifiers (i.e. `twit:username` and `insta:username` would lead to the same pepper for `username`). One way of getting around this would be treating phone numbers as special cases (i.e. no prefix in the pepper generation) but including the prefix for other blinded messages, though I personally would want to get additional eyes on this from the crypto team folks about the security implications of this before moving forward with this. Let me know if I'm misunderstanding something about this though!!

**Comment by eelanagaraj** [comment]
more generally, in addition to the unit tests we should ideally sanity-check this by manually testing that we get the same pepper/identifier for a phone number with the current version of the SDK & updated version from this PR. Reason being is that a breaking change (i.e. we are missing a '+' sign somewhere or added a space, or whatever) would be really hard to spot IMO and would be very difficult to debug!

File: packages/sdk/identity/src/odis/identifier.ts
Comment (isabellewei) [reply packages/sdk/identity/src/odis/identifier.ts]: using `plaintextIdentifier` -> `blindedIdentifier` -> `obfuscatedIdentifier` instead

File: packages/sdk/identity/src/odis/identifier.ts
Comment (isabellewei) [reply packages/sdk/identity/src/odis/identifier.ts]: I was hesitant to change the blinding because I wanted to make sure we are still interoperable with pre-existing phone number identifiers. I'm worried handling the phone numbers differently might be confusing, especially if someone chooses to forgo the sdk and query ODIS/blind/hash themselves.

But definitely still an important point, thanks for pointing it out! How much of a security risk do we think it is for the same plaintext identifier across different types to have the same pepper? @nategraf

File: packages/sdk/identity/src/odis/identifier.ts
Comment (alecps) [inline packages/sdk/identity/src/odis/identifier.ts]: Why not make this just an enum of strings rather than having this enum plus the switch statement below?

File: packages/sdk/identity/src/odis/identifier.ts
Comment (alecps) [inline packages/sdk/identity/src/odis/identifier.ts:46]: This would be a good place to add some comments clarifying what each of these fields actually is

File: packages/sdk/identity/src/odis/identifier.ts
Comment (alecps) [inline packages/sdk/identity/src/odis/identifier.ts]: Could just pass in `identifierPrefix`

File: packages/sdk/identity/src/odis/identifier.ts
Comment (alecps) [inline packages/sdk/identity/src/odis/identifier.ts:65]: if time permits, it might be good to add docstrings for exported sdk functions

File: packages/sdk/identity/src/odis/identifier.ts
Comment (alecps) [inline packages/sdk/identity/src/odis/identifier.ts]:
```suggestion
 * Response can be passed into getObfuscatedIdentifierFromSignature
```

File: packages/sdk/identity/src/odis/identifier.ts
Comment (alecps) [inline packages/sdk/identity/src/odis/identifier.ts]: I think this would be simplified a bit if we only ever passed around the prefix as an enum member per my suggestions above

File: packages/sdk/identity/src/odis/identifier.ts
Comment (alecps) [inline packages/sdk/identity/src/odis/identifier.ts:223]: Do you need to cast this here? It looks like soliditySha3 already returns a string.

File: packages/sdk/identity/src/odis/phone-number-identifier.ts
Comment (alecps) [inline packages/sdk/identity/src/odis/phone-number-identifier.ts]: This PEPPER_CHAR_LENGTH constant is defined twice, once here and once in the new identifier.ts file

File: packages/sdk/identity/src/odis/phone-number-identifier.ts
Comment (alecps) [inline packages/sdk/identity/src/odis/phone-number-identifier.ts]: This is also duplicated in both files

File: packages/sdk/identity/src/odis/identifier.test.ts
Comment (alecps) [inline packages/sdk/identity/src/odis/identifier.test.ts]:
```suggestion
    it('Preblinding the off-chain identifier', async () => {
```

File: packages/sdk/identity/src/odis/identifier.ts
Comment (alecps) [reply packages/sdk/identity/src/odis/identifier.ts]: Doing it this way basically means that if you query once for `username`, you can derive the on-chain identifier for twitter://username  and instagram://username without querying a second time

File: packages/sdk/identity/src/odis/identifier.ts
Comment (alecps) [inline packages/sdk/identity/src/odis/identifier.ts]: It would be good to include a link to the resources used to determine that these are appropriate prefixes

**Review by alecps — COMMENTED** [review]
The two main things I see that need addressing before we can merge this are the same as Eela pointed out. 
1. We need a backwards compatibility check somewhere to ensure that we're getting the same identifiers as before. This could maybe be added to the combiner e2e tests 
2. It might be better if different prefixes yielded different peppers. If we decide to keep things as they are, we should be sure to document our reasoning.

Though it would be amazing to squeeze this feature into this round of releases, there's a part of me that feels a bit uneasy about rushing this and adding new complexity right at the finish line. Normally this type of feature would be given much more time for careful design / review, and I wonder if it might be best to adjust our expectations to getting this merged after we've released ASv2. I'm not saying we can't get it in beforehand, but we should step back and weigh both options

File: packages/sdk/identity/src/odis/identifier.ts
Comment (isabellewei) [reply packages/sdk/identity/src/odis/identifier.ts]: I copied this structure over from existing code, but this is a good point!

File: packages/sdk/identity/src/odis/identifier.ts
Comment (isabellewei) [reply packages/sdk/identity/src/odis/identifier.ts:223]: it returns `string || null`, so we need to force it to be a `string` to make typescript happy

File: packages/sdk/identity/src/odis/phone-number-identifier.ts
Comment (isabellewei) [reply packages/sdk/identity/src/odis/phone-number-identifier.ts]: Ideally we can remove these from this file, but it's being imported in the `attestation-service` package. I've left it here to not break any builds. We can remove when removing rest of ASv1 code

**Comment by isabellewei** [comment]
After chatting with @nategraf, agreed that it's best to also append prefixes **before** blinding the identifiers, but skip this step for phone numbers to maintain backwards compatibility.

I've also added a backwards compatibility test for phone numbers that import an old version of the `@celo/identity` sdk to compare against the new changes

File: packages/sdk/identity/src/odis/identifier.ts
Comment (isabellewei) [inline packages/sdk/identity/src/odis/identifier.ts:52]: made the `pepper` parameter mandatory

File: packages/sdk/identity/package.json
Comment (aaronmgdr) [inline packages/sdk/identity/package.json]: ah it depends on itself?

Comment (aaronmgdr) [reply packages/sdk/identity/package.json]: oh for the test

Comment (aaronmgdr) [reply packages/sdk/identity/package.json]: hmm it would probably be smart to alias this

Comment (alecps) [reply packages/sdk/identity/package.json]: +1 this should be aliased https://www.dev-diaries.com/social-posts/install-npm-package-under-alias/

File: packages/sdk/identity/src/odis/phone-number-identifier.ts
Comment (alecps) [reply packages/sdk/identity/src/odis/phone-number-identifier.ts]: I'd vote to update the imports in the attestation service or make the attestation service rely on an older version of the identity package rather than publish this package with the same function defined in two places

Comment (alecps) [reply packages/sdk/identity/src/odis/phone-number-identifier.ts]: Since we're releasing a major version of the identity package, we can make this breaking change

File: packages/sdk/identity/src/odis/identifier.ts
Comment (alecps) [reply packages/sdk/identity/src/odis/identifier.ts:46]: The comments added here are good. It would be better to also explain how they're derived from each other. I.e. 
- obfuscatedIdentifier is obtained by hashing (plaintext identifier + pepper) and truncating
- pepper is obtained by hashing the unblindedSignature

File: packages/sdk/identity/src/odis/identifier.ts
Comment (alecps) [inline packages/sdk/identity/src/odis/identifier.ts]: This should use the enum type

Comment (alecps) [reply packages/sdk/identity/src/odis/identifier.ts]: so that the user can't just pass in any string

File: packages/sdk/identity/src/odis/identifier.ts
Comment (alecps) [inline packages/sdk/identity/src/odis/identifier.ts:133]: could you add a comment explaining why this is a special case?

Comment (alecps) [reply packages/sdk/identity/src/odis/identifier.ts:133]: Also I would personally vote to move this check into getPrefixedIdentifier

File: packages/sdk/identity/src/odis/identifier.ts
Comment (alecps) [inline packages/sdk/identity/src/odis/identifier.ts:65]: much of this logic is duplicated in getPhoneNumberIdentifier

Comment (alecps) [reply packages/sdk/identity/src/odis/identifier.ts:65]: Ideally getPhoneNumberIdentifier would just wrap this completely

File: packages/sdk/identity/src/odis/identifier.ts
Comment (alecps) [inline packages/sdk/identity/src/odis/identifier.ts]: should just be IdentifierPrefix, right?

File: packages/sdk/identity/src/odis/identifier.ts
Comment (alecps) [inline packages/sdk/identity/src/odis/identifier.ts]: this should use the enum type

File: packages/sdk/identity/src/odis/identifier.ts
Comment (alecps) [inline packages/sdk/identity/src/odis/identifier.ts]: please use enum type

**Review by alecps — CHANGES_REQUESTED** [review]
looking good so far!

File: packages/sdk/identity/src/odis/identifier.ts
Comment (aaronmgdr) [reply packages/sdk/identity/src/odis/identifier.ts]: +1

File: packages/sdk/identity/package.json
Comment (isabellewei) [reply packages/sdk/identity/package.json]: good suggestion thanks!

File: packages/sdk/identity/src/odis/identifier.ts
Comment (isabellewei) [reply packages/sdk/identity/src/odis/identifier.ts]: Do we want to force the user to use one of the predefined prefixes? I think we should allow them to pass in an arbitrary string if they want to

File: packages/sdk/identity/src/odis/phone-number-identifier.ts
Comment (isabellewei) [reply packages/sdk/identity/src/odis/phone-number-identifier.ts]: If we're ok with making breaking changes, we could even just get rid of the entire`phone-number-identifier` file and get users to switch to the functions in `identifier`

File: packages/sdk/identity/src/odis/identifier.ts
Comment (isabellewei) [reply packages/sdk/identity/src/odis/identifier.ts:133]: `getPrefixedIdentifier` is also used when constructing the final hash, where there is no special handling for phone numbers

File: packages/sdk/identity/src/odis/identifier-backwards-compatibility.test.ts
Comment (alecps) [inline packages/sdk/identity/src/odis/identifier-backwards-compatibility.test.ts]: this should be aliased to make it clear that it's importing from an old version of this package

File: packages/sdk/identity/src/odis/identifier-backwards-compatibility.test.ts
Comment (alecps) [inline packages/sdk/identity/src/odis/identifier-backwards-compatibility.test.ts:61]: it might be good to also have a test that verifies that the outputs are different if an alternate prefix is passed in

File: packages/sdk/identity/src/odis/identifier.ts
Comment (alecps) [reply packages/sdk/identity/src/odis/identifier.ts]: hmm I mean it leaves a sharper edge for devs to cut themselves on if they can pass in an arbitrary string, and might make them less inclined to put up a PR adding to the prefix list. That being said, it is a big lift to update and release the sdk to add a prefix. I think right now if we require they pass in a predefined prefix devs can get around it by casting 'prefix as IdentifierPrefix'. Needing to keep this eyesore in their code may help flag that the prefix has not yet been added to the sdk and prompt devs to put up a PR. Happy to defer to other reviewers here tho @eelanagaraj @nategraf

File: packages/sdk/identity/src/odis/identifier-backwards-compatibility.test.ts
Comment (alecps) [inline packages/sdk/identity/src/odis/identifier-backwards-compatibility.test.ts:1]: this test should be ignored in CI since they'll be prone to flakiness. We do this in the odis packages by using the '--testPathIgnorePatterns test/end-to-end' flag

File: packages/sdk/identity/src/odis/identifier-backwards-compatibility.test.ts
Comment (alecps) [inline packages/sdk/identity/src/odis/identifier-backwards-compatibility.test.ts]: can you hardcode the seed and expected result so it's easier to verify the tests work as expected without actually running them?

File: packages/sdk/identity/src/odis/identifier.ts
Comment (isabellewei) [reply packages/sdk/identity/src/odis/identifier.ts]: convo continued in https://github.com/celo-org/celo-monorepo/pull/9985#discussion_r1030759270

File: packages/sdk/identity/src/odis/identifier.ts
Comment (isabellewei) [reply packages/sdk/identity/src/odis/identifier.ts]: convo continued in https://github.com/celo-org/celo-monorepo/pull/9985#discussion_r1030759270

File: packages/sdk/identity/src/odis/identifier.ts
Comment (isabellewei) [reply packages/sdk/identity/src/odis/identifier.ts]: convo continued in https://github.com/celo-org/celo-monorepo/pull/9985#discussion_r1030759270

File: packages/sdk/identity/src/odis/identifier-backwards-compatibility.test.ts
Comment (isabellewei) [reply packages/sdk/identity/src/odis/identifier-backwards-compatibility.test.ts]: do you mean to compare the results against hardcoded identifiers/peppers?

File: packages/sdk/identity/src/odis/phone-number-identifier.ts
Comment (isabellewei) [reply packages/sdk/identity/src/odis/phone-number-identifier.ts]: https://github.com/celo-org/celo-monorepo/pull/9985#discussion_r1010487322

File: packages/sdk/identity/src/odis/phone-number-identifier.ts
Comment (isabellewei) [reply packages/sdk/identity/src/odis/phone-number-identifier.ts]: after sync discussion: I will set the `attestation-service` package to depend on a specific version of the identity sdk to avoid breaking changes

File: packages/sdk/identity/src/odis/identifier-backwards-compatibility.test.ts
Comment (nategraf) [inline packages/sdk/identity/src/odis/identifier-backwards-compatibility.test.ts:15]: Are these comments intentionally left in the code?

File: packages/attestation-service/package.json
Comment (nategraf) [inline packages/attestation-service/package.json]: Is this older, pre-2.0, version the correct version to be using?

File: packages/sdk/identity/src/odis/identifier.ts
Comment (nategraf) [inline packages/sdk/identity/src/odis/identifier.ts]: Putting this into a [TSDoc](https://tsdoc.org/) comment will allow it to appear in the online docs.
```suggestion
/** Details with the private plaintext identifier and obfuscated identifier, which can be made public.
 * plaintext identifier: off-chain information, ex: phone number, twitter handle, email, etc.
 * blinded identifier: obtained by blinding the plaintext identifier
 * blinded signature: blinded identifier signed by ODIS
 * unblinded signatue: obtained by unblinding the blinded signature
 * pepper: unique secret, obtained by hashing the unblinded signature
 * obfuscated identifier: identifier used for on-chain attestations, obtained by hashing the plaintext identifier and pepper
 */
```

File: packages/sdk/identity/src/odis/identifier.ts
Comment (nategraf) [inline packages/sdk/identity/src/odis/identifier.ts]:
```suggestion
```

File: packages/sdk/identity/src/odis/identifier.ts
Comment (nategraf) [inline packages/sdk/identity/src/odis/identifier.ts:59]: Suggest a note to call out that this function makes a network call and consumes quota.
```suggestion
 * Performs blinding, querying, and unblinding
 *
 * This function will send a request to ODIS, authorized by the provided signer.
 * This method consumes ODIS quota on the account provided by the signer.
```

File: packages/sdk/identity/src/odis/identifier.ts
Comment (nategraf) [inline packages/sdk/identity/src/odis/identifier.ts]: It's possible for many identifiers to contain underscores (e.g. Twitter) which raises the possibility of domain collisions. Let's use a simply trick, which is to hash the identifier, then tack on the pepper to the hash and hash the whole thing.
```suggestion
  const value = (sha3(getPrefixedIdentifier(plaintextIdentifier, identifierPrefix)) as string) + PEPPER_SEPARATOR + pepper
  return sha3(value) as string
```

**Review by nategraf — APPROVED** [review]
Made a few comments, one with relation to the hashing of identifiers to suggest a two step hash. Overall it looks good.

File: packages/sdk/identity/src/odis/identifier-backwards-compatibility.test.ts
Comment (isabellewei) [reply packages/sdk/identity/src/odis/identifier-backwards-compatibility.test.ts:15]: yes, in case the information is needed for writing further tests

File: packages/sdk/identity/src/odis/identifier.ts
Comment (isabellewei) [reply packages/sdk/identity/src/odis/identifier.ts]: good suggestion thanks! I've committed this change but I think they are actually planning on completely taking down the sdk docs site

File: packages/sdk/identity/src/odis/identifier.ts
Comment (isabellewei) [reply packages/sdk/identity/src/odis/identifier.ts]: This breaks backwards compatibility :(

Comment (isabellewei) [reply packages/sdk/identity/src/odis/identifier.ts]: putting this modification behind a conditional statement checking if the identifier is a phone number, similar to what we're doing for blinding

File: packages/attestation-service/package.json
Comment (isabellewei) [reply packages/attestation-service/package.json]: yes! It breaks with newer versions

File: packages/sdk/identity/src/odis/identifier.ts
Comment (nategraf) [reply packages/sdk/identity/src/odis/identifier.ts]: Looking at this now, I think it' ok and probably preferred to let devs specify this prefix as they wish. There are some edge cases that could cause ambiguity (e.g. between { prefix: "url", identifier "https://exmaple.com" } and { prefix: "url://https", identifier: "example.com" }) but not ones you would hit without really going out of your way.

File: packages/sdk/identity/src/odis/identifier-backwards-compatibility.test.ts
Comment (alecps) [inline packages/sdk/identity/src/odis/identifier-backwards-compatibility.test.ts:19]: why is this commented out public key the same as the private key above?

File: packages/sdk/identity/src/odis/identifier.ts
Comment (alecps) [inline packages/sdk/identity/src/odis/identifier.ts:2]: we're gonna need to add this to OdisUtils in the index file

File: packages/sdk/identity/src/odis/identifier-backwards-compatibility.test.ts
Comment (alecps) [inline packages/sdk/identity/src/odis/identifier-backwards-compatibility.test.ts:107]: does ts not complain about this?

**Review by alecps — APPROVED** [review]
Nice work!

File: packages/sdk/identity/src/odis/identifier-backwards-compatibility.test.ts
Comment (alecps) [reply packages/sdk/identity/src/odis/identifier-backwards-compatibility.test.ts:107]: since it isn't of the enum type?

**Pull request merged**
